cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/Internal.hs

```diff
@@ -288,6 +307,22 @@ withNewAffine_ = fmap fst . withNewAffine
 withNewAffine' :: BLS curve => (AffinePtr curve -> IO a) -> IO (Affine curve)
 withNewAffine' = fmap snd . withNewAffine
 
+withAffineVector :: NonEmpty.NonEmpty (Affine curve) -> (AffinePtrVector curve -> IO a) -> IO a
+withAffineVector affines go = do
+  let numAffines = NonEmpty.length affines
+      sizeReference = sizeOf (undefined :: Ptr ())
+  allocaBytes (numAffines * sizeReference) $ \ptr ->
+    -- The accumulate function ensures that each `withAffine` call is properly nested.
```

Comment:
I know very little about the FFI and this problem certainly wouldn't have occurred to me.  I found myself worrying about laziness though, and whether there should be some `!`s somewhere.  Does the monadic code ensure that all of this happens in the right order?